libsolidity/interface/CompilerStack.cpp

```diff
@@ -2010,18 +2010,47 @@ experimental::Analysis const& CompilerStack::experimentalAnalysis() const
 	return *m_experimentalAnalysis;
 }
 
-void CompilerStack::reportUnimplementedFeatureError(UnimplementedFeatureError const& _error)
+void CompilerStack::reportUnimplementedFeatureError(
+	UnimplementedFeatureError const& _error,
+	ContractDefinition const* _contractDefinition
+)
 {
 	solAssert(_error.comment(), "Errors must include a message for the user.");
+	if (_error.sourceLocation().sourceName)
+		solAssert(m_sources.count(*_error.sourceLocation().sourceName) != 0);
+
+	m_errorReporter.unimplementedFeatureError(
+		1834_error,
+		(_error.sourceLocation().sourceName || !_contractDefinition) ?
+			_error.sourceLocation() :
+			_contractDefinition->location(),
+		*_error.comment()
+	);
+}
 
-	m_errorReporter.unimplementedFeatureError(1834_error, _error.sourceLocation(), *_error.comment());
+void CompilerStack::reportCodeGenerationError(Error const& _error, ContractDefinition const* _contractDefinition)
+{
+	solAssert(_error.type() == Error::Type::CodeGenerationError);
+	solAssert(_error.comment(), "Errors must include a message for the user.");
+	if (_error.sourceLocation() && _error.sourceLocation()->sourceName)
+		solAssert(m_sources.count(*_error.sourceLocation()->sourceName) != 0);
+	solAssert(_contractDefinition);
```

Comment:
`reportUnimplementedFeatureError` is also used in places where we don't necessarily know the contract. I.e. analysis, which runs each stage across all contracts rather than doing everything for one contract and proceeding to the next one.

I was actually a bit conflicted about this, but in the end I went for the pointer because of that uniformity and because this way I also don't have to repeat the assert at every use (usually I have a pointer anyway).

Reply:
Yeah it makes sense, thank you for clarification! I think uniformity is good in this case and if we ever need it, the implementation can be extended to cater the `nullptr` case too.